### PR TITLE
[Give Ratings] API implementation

### LINKED ISSUE #1884: Submit Rating

### PR DESCRIPTION
**Pull request by danielebogo**
Fixes #1884

This PR implement the 2 APIs toget and submit a podcast rating. The APIs are:
• `/user/podcast_rating/add` to update a podcast that has previous rate
• `/user/podcast_rating/show` to fetch the individual rate
The APIs are implement using 2 new tasks based on the Protobuf implementation.
This PR also adds
• hide the Submit button when the the previous rate and current rate selected to update is the same
• update the dismiss button position based on Figma design

## To test

> [!NOTE]
> Have `giveRatings` feature flag enabled
> Please use Staging to test

1. Open a podcast that you did not listen any episode
2. Try to rate a podcast that you did not rate before
3. ✅ Ensure the rating screen is opened with submit button disabled
4. Select any rate
5. ✅ Ensure the submit button is enabled
6. Rate the podcast
7. Open the rating screen again
8. ✅ Ensure you see the rate you gave in step 6 and submit button is hidden
9. Select a different rate
10. ✅ Ensure submit button is visible
11. Disable the feature flag
12. ✅ Test everything still works as normal

## Recording

https://github.com/Automattic/pocket-casts-ios/assets/912252/fa70d2c7-a17e-46a2-82a2-73d66c71fcbb

## Checklist

- [ ] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [ ] I have considered adding unit tests for my changes.
- [ ] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Modules/Server/Sources/PocketCastsServer/Private/API Tasks/UserPodcastRatingTask.swift [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+            if response == nil {
+                completion?(false)
+                return
+            }
+
+            if httpStatus == ServerConstants.HttpConstants.ok {
+                FileLog.shared.addMessage("Add rating success for podcast \(uuid)")
+            } else {
+                FileLog.shared.addMessage("Failed to add rating for podcast \(uuid)")
```
Comment (leandroalonso) [inline Modules/Server/Sources/PocketCastsServer/Private/API Tasks/UserPodcastRatingTask.swift]: Maybe log the returned `httpStatus`?

File: Modules/Server/Sources/PocketCastsServer/Public/API/ApiServerHandler+ UserPodcastRating.swift [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import Foundation
+
+public extension ApiServerHandler {
```
Comment (leandroalonso) [inline Modules/Server/Sources/PocketCastsServer/Public/API/ApiServerHandler+ UserPodcastRating.swift]: The name of the file has an empty space after `+`, not sure if this is intentional?

File: Modules/Server/Sources/PocketCastsServer/Public/Helpers/JSONDecodableURLTask.swift [hunk 3]
```diff
@@ -23,21 +23,43 @@ public struct JSONDecodableURLTask<Response: Decodable> {
...
+            throw JSONDecodableURLTaskError.ratingForbidden
+        default:
+            break
+        }
+    }
+
+    enum JSONDecodableURLTaskError: Error {
+        case notValidResponse
+        case objectNotFound
+        case ratingForbidden
```
Comment (leandroalonso) [inline Modules/Server/Sources/PocketCastsServer/Public/Helpers/JSONDecodableURLTask.swift]: This is a generic class, shall we name this just `forbidden` without the `rating`?

**Review by leandroalonso — APPROVED** [review]
✅ Tested the mentioned steps
✅ Tested while logged out (the behavior now is weird but I see it will be addressed in a future task)
✅ Tested removing the app, installing, logging in again

**Pull request merged**